### ISSUE TITLE
The future of utilities

### ISSUE THREAD

**Issue by florianbouvot** [inuitcss/inuitcss#396]
My (personal) vision about utilities:

- Responsive : every utilities must have responsive variants by default _(with a way to toggle off)_
- Customizable : set easily your own values using (simple) sass map
- Simple code base (empowering users)

Minimum set:
- Display
- Flexbox
- Font-size (headings)
- Spacing (padding, margin)
- Sizing (width)

Why not set:
- Float (someone use it?)
- Text alignement
- Text transform 
- Font weight

I will add a quick overview and the discussion should begin 😉

**Comment by LayZeeDK** [comment]
Just don't overdo it like [Tailwind CSS](https://tailwindcss.com/docs/what-is-tailwind/).
![image](https://user-images.githubusercontent.com/6364586/47009985-21337c80-d13e-11e8-9d62-b5886ff3b9ff.png)

**Comment by florianbouvot** [comment]
I add a first example with to kind of utilities:
- `utilities.display`: generated utilities with responsive variants
- `utilities.margin`: generated utilities with customizable values and responsive variants _(sorry for short classname)_

I think it is very simple and not over engineered: yes there are some downside 🤫 (and please don't tell me file size because GZIP 😉).

**Comment by florianbouvot** [comment]
> Just don't overdo it like [Tailwind CSS](https://tailwindcss.com/docs/what-is-tailwind/).

I hear 😉 Inuitcss are not a utility library but in older version we've more utilities and it was pretty cool too 😎.

**Comment by herzinger** [comment]
I'm against, to be honest. I'm extremely against the "atomic design" paradigm, and I find that an over abundance of utilities breeds lazy development and design inconsistences, so I avoid most of them in my projects.

Of course, inuitcss is modular enough that someone with my view can simply not import their files, but their existence in the project incentives usage, like it's endorsing it as a good practice.

**Comment by florianbouvot** [comment]
@herzinger I understand your opinion, I will wait for more feedback before working on utilities part.

**Comment by lhermann** [comment]
@florianbouvot I support your suggestion.

There are some utilities we find ourselves including in every project anyway, why not add it by default?

The ones I really miss currently are:
- Display: `block`, `inline-block` and `inline`
- Text alignment: `left`, `right` and `center`
- Text Size: I am usually defining `tiny`, `small`, `default`, `large` and `huge`

For float, text transform and font weight (since we got `<strong>`) I don't see a big need.

**Comment by herzinger** [comment]
Text size utilities are waaay too optionated. For one, they shouldn't really follow the global spacing unit, so this naming convention would just breed confusion. The closest to that I use are the headings variables/utilities. I use the greek letters convention, but theres nothing wrong with inuitcss's default `h#`.

Don't get me wrong, I can see why you would want to define font sizes, but 1. I'll argue they should be variables or a map, not utilities, and 2. This is a per project design decision and inuitcss shouldn't touch it.

**Comment by herzinger** [comment]
@LayZeeDK tailwind is a "utilities first" framework built with the atomic design paradigm in mind. It's not that they overdid it: mostly not writing new classes is a principle of their methodology (which I'm REALLY against, by the way), as much as our layers separation is a principle of the ITCSS we follow.

**Comment by LayZeeDK** [comment]
Sorry everyone for going a bit off-topic here, but to reply to @herzinger:
It was of course a joke when I wrote that Tailwind overdid it :-) Personally, I am a fan of keeping markup in HTML, styles in CSS and behaviour in JavaScript, so when I see that many inline styles disguised as HTML classes, it seems wrong to me. The markup is so full of styles that it becomes a distraction from the content instead of a help to discover the visual semantics.

It seems that some developers like to have styles, markup, rendering logic, business logic, and behaviour in the same place. I am definitely not one of those. I tried using Angular Flex-Layout which inlines styles and breakpoint styles for you based on Angular's declarative template syntax - much like HTML classes, but really it is dynamic styles in inlined JavaScript. It does not do anything that you cannot do with basic CSS. It is not much more than flexbox in your template. Semantic classes such as the BEM, OOCSS and SMACSS naming conventions are far more helpful and in the true spirit of HTML and CSS as far as I am concerned.

**Comment by lhermann** [comment]
> Don't get me wrong, I can see why you would want to define font sizes, but 1. I'll argue they should be variables or a map, not utilities, and 2. This is a per project design decision and inuitcss shouldn't touch it.

@herzinger your argument is valid. Personally I would wish there were such variables predefined, even if I have to change them per project.

For example, I first tried `em` and `%` units (like default `1em` and small `0.8em`) but it got messy. So now I take the `$inuit-global-font-size: 16px` and make myself a `...-small: 14px`, `...-tiny: 12px`, just going up and down 2px for each level which worked really well. But is it the best way? Therefore I thought a default might be nice.

**Comment by csshugs** [comment]
With the development of version 6 of inuitcss, we were very strict on removing stuff that is not totally essential and opinionated for everyone using the framework. I’m still and more than ever convinced that this should stay our principle (and maybe we should write that down somewhere): Just provide a basic set of objects and utilities that are not opinionated so really every user can benefit and there is nothing ‘needless’ in the framework that is not used. Everything else should be added on project-level or as additional inuitcss-plugins/addons.

I think the utilities we currently have are good, and there may be other utilities that could suit the framework. But I’m afraid that the above mentioned suggestions are too specific/opinionated/needless. 

It seems like we emptied our tool shed in order to tidy up everything in version 6, and in version 7 we cram everything back in.

Optimising our current existing utilities is totally fine, but adding all those utilities again is not the right way to go for me.

**Comment by andrejilderda** [comment]
I agree with @csshugs, but I think @lhermann has a point. I think one of the nice things about Inuit is how strict the conventions being used are. I think lhermann's question is, how to solve responsive typography the Inuit way? 

This is indeed more opinionated and I prefer this to be available as a plugin in stead of adding it to the core. However, a list of (unofficial) Inuitcss-plugins/addons is currently missing. Is it an idea to add this to the readme or start some sort of 'Awesome' Inuit-list? 
I believe this motivates users to create plugins, helps people understand the Inuit philosophy and make Inuit even more useful than it already is.

**Comment by herzinger** [comment]
> For example, I first tried em and % units (like default 1em and small 0.8em) but it got messy. 

The `@inuit-font-size` mixin receives px values and handles rem conversion, so storing px values is the way to go but...

> So now I take the $inuit-global-font-size: 16px and make myself a ...-small: 14px, ...-tiny: 12px, just going up and down 2px for each level which worked really well. But is it the best way? Therefore I thought a default might be nice.

The `$inuit-global-font-size` is 16px because that is conventionally the browsers base font size. There's not, and there can't be a convention for topography sizes. Even the fibonacci based default for headings is really opinionated, and I have expressed more than once how it's the most uncharacteristically opinionated piece of code in the framework. Typography is 100% a design choice, and should be handled aways in a per project basis. There's no sense in inuitcss defining a pattern for that, because each and every project should have their own values. So no, I don't believe this is a the best way. 

>  I think lhermann's question is, how to solve responsive typography the Inuit way?

Responsive typography isn't even the same topic, but the same concept applies. I for one don't even use any kind of complex system for that, just regular old media-queries in the elements and or components layers, wherever it's necessary. So there's no mechanic pattern, and it's also 100% a design decision made per project.

**Comment by csshugs** [comment]
So, after thinking about this (for over a year now 😜), I really think that inuitcss should distance itself from utility-first frameworks such as Tailwind CSS. Not that they are generally bad or promoting bad practice. Not at all! If that’s how some person wants to build their UI, that’s totally fine. 

**BUT:** That’s just not the way inuitcss wants to approach things. If you like the utility-driven approach, go with Tailwind CSS or any other framework out there. But if you are more the component-/object-driven type of person, inuitcss is well suited for you. With this philosophy in mind, there is no need to make any compromises by providing just a few helpful utilities. Sure, we currently have a minimal set of utilities that are critical and won’t be deleted. But if you find, some utilities are currently missing for your project(s), the same basic principle applies as it does for the entire framework: “Just add it yourself”. inuitcss doesn’t want to be that big framework that comes with all this optional stuff where I have to handpick the 10% I really need. We rather want every inuitcss user to use 80%–100% of the framework in each of their projects. This way, they have the freedom to add the real custom stuff which is specific to each project theirselves.

**To summarize:**
* If you like utility-driven development, maybe a framework like Tailwind CSS is better suited for you.
* If you just need a few extra utilities, add them yourself—maybe provide them as inuitcss plugins/extensions for other inuitcss users to use.
* We won’t add more utilities to inuitcss, unless they are absolutely needed for the essential working of the framework.
